pkg/provisioner/ironic/ironic.go

```diff
@@ -1734,46 +1734,41 @@ func (p *ironicProvisioner) PowerOn() (result provisioner.Result, err error) {
 			p.log.Info("waiting for power status to change")
 			return operationContinuing(powerRequeueDelay)
 		}
-		result, err = p.changePower(ironicNode, nodes.PowerOn)
-		switch err.(type) {
-		case nil:
-		case HostLockedError:
-		default:
-			return transientError(errors.Wrap(err, "failed to power on host"))
+		if ironicNode.LastError != "" && !force {
+			p.log.Info("PowerOn operation failed", "msg", ironicNode.LastError)
+			return operationFailed(fmt.Sprintf("PowerOn operation failed: %s",
+				ironicNode.LastError))
+		}
+		if result, err = p.changePower(ironicNode, nodes.PowerOn); err == nil {
+			p.publisher("PowerOn", "Host powered on")
+			return result, nil
+		} else {
+			switch err.(type) {
+			case HostLockedError:
+				return retryAfterDelay(powerRequeueDelay)
+			default:
+				return transientError(errors.Wrap(err, "failed to PowerOn node"))
+			}
 		}
-		p.publisher("PowerOn", "Host powered on")
 	}
-
 	return result, nil
 }
 
 // PowerOff ensures the server is powered off independently of any image
 // provisioning operation.
-func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode) (result provisioner.Result, err error) {
+func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode, force bool) (result provisioner.Result, err error) {
 	p.log.Info(fmt.Sprintf("ensuring host is powered off (mode: %s)", rebootMode))
 
-	if rebootMode == metal3v1alpha1.RebootModeHard {
-		result, err = p.hardPowerOff()
+	if rebootMode == metal3v1alpha1.RebootModeSoft {
+		return p.softPowerOff()
```

Comment:
It's a little unfortunate that we have to record a user-visible failure here when soft power off isn't supported. This is still an improvement to the readability though, because we're no longer calling `softPowerOff()` to check the state of a hard power off we fell back to.
I think it should be possible to get the best of both in future with some refactoring...